File: typescript/packages/docker/src/exec.ts
import type { Sinks, Started } from "@threads/core/adapter";
import { z } from "zod";
import { type Engine, failure } from "./engine";
import { parsed, unavailable } from "./wire";

// One exec through the Engine API: the attach reply is Docker's multiplexed stream, whose
// 8-byte frame headers are parsed strictly (container output is a trust boundary, so an
// unknown stream byte or an absurd length is a typed failure, never a guess). The exit code
// comes from the exec's own record, after the stream ends.

const ExecCreated = z.object({ Id: z.string().min(1) });
const ExecInspected = z.object({ ExitCode: z.int().nullish() });

/** The most bytes one frame may claim: beyond this the stream is not Docker's. */
const MAX_FRAME = 64 * 2 ** 20;
const HEADER = 8;
/** How often, and how long apart, an exit code the daemon hasn't stored yet is looked for. */
const SETTLE_LOOKS = 5;
const SETTLE_MS = 20;

export type ExecSpec = {
  /** "0" for the supervisor and its probes, "1000" for the kit's own scripts. */
  readonly user: string;
  readonly cmd: readonly string[];
};

/** Feeds Docker's multiplexed stream into the kit's sinks, one frame at a time. */
/** One frame header: `[stream, 0, 0, 0, length:u32be]`, or a typed failure. */
function headerAt(held: Uint8Array): {
  readonly stream: 1 | 2;
  readonly length: number;
} {
  const stream = held[0];
  if (stream !== 1 && stream !== 2)
    throw unavailable(
      `Docker's output stream carries an unknown frame type ${stream}`,
    );
  const length = new DataView(held.buffer, held.byteOffset + 4, 4).getUint32(0);
  if (length > MAX_FRAME)
    throw unavailable(
      `Docker's output stream claims a frame of ${length} bytes`,
    );
  return { stream, length };
}

export function demux(sinks: Sinks): {
  readonly push: (chunk: Uint8Array) => void;
  readonly end: () => void;
} {
  let held = new Uint8Array(0);
  return {
    push: (chunk) => {
      const joined = new Uint8Array(held.length + chunk.length);
      joined.set(held);
      joined.set(chunk, held.length);
      held = joined;
      while (held.length >= HEADER) {
        const { stream, length } = headerAt(held);
        if (held.length < HEADER + length) return;
        const body = held.slice(HEADER, HEADER + length);
        held = held.slice(HEADER + length);
        if (stream === 1) sinks.stdout(body);
        else sinks.stderr(body);
      }
    },
    end: () => {
      if (held.length > 0)
        throw unavailable("Docker's output stream ends inside a frame");
    },
  };
}

/** Creates the exec, starting the container first when a stopped one refused it. */
async function created(
  engine: Engine,
  name: string,
  spec: ExecSpec,
): Promise<string> {
  const body = {
    json: {
      AttachStdout: true,
      AttachStderr: true,
      AttachStdin: false,
      Tty: false,
      User: spec.user,
      Env: [],
      WorkingDir: "/workspace",
      Cmd: spec.cmd,
    },
  };
  const at = `/containers/${encodeURIComponent(name)}/exec`;
  const first = await engine.send("POST", at, body);
  if (first.status !== 409) {
    if (!first.ok) throw await failure(first, "an exec create");
    return parsed(ExecCreated, await first.text(), "an exec create").Id;
  }
  // A stopped container is started again by the next exec; attach never starts one.
  await first.text();
  const started = await engine.send(
    "POST",
    `/containers/${encodeURIComponent(name)}/start`,
  );
  if (!started.ok && started.status !== 304)
    throw await failure(started, "a container start");
  await started.text();
  return (await engine.json(ExecCreated, "an exec create", "POST", at, body))
    .Id;
}

/**
 * The exec's own exit status. The daemon can still be storing it when the output stream
 * closes, so a null is looked at again a few times before it is taken as an answer.
 */
async function exitCode(engine: Engine, id: string): Promise<number | null> {
  for (let look = 0; ; look += 1) {
    const res = await engine.send(
      "GET",
      `/exec/${encodeURIComponent(id)}/json`,
    );
    if (!res.ok) throw await failure(res, "an exec inspect");
    const code = parsed(
      ExecInspected,
      await res.text(),
      "an exec inspect",
    ).ExitCode;
    if (code !== undefined && code !== null) return code;
    if (look >= SETTLE_LOOKS) return null;
    await new Promise((done) => setTimeout(done, SETTLE_MS));
  }
}

/**
 * Starts `spec` in the container and streams its output into `sinks`. Resolves once Docker
 * accepted the attach; `exit` resolves with the exec's code once the stream has ended.
 */
export async function startExec(
  engine: Engine,
  name: string,
  spec: ExecSpec,
  sinks: Sinks,
  /** What an exec that ends with no code means; a terminate probe reads it as its own row. */
  noCode: () => number,
): Promise<Started> {
  const id = await created(engine, name, spec);
  const res = await engine.send(
    "POST",
    `/exec/${encodeURIComponent(id)}/start`,
    {
      json: { Detach: false, Tty: false },
    },
  );
  if (!res.ok) throw await failure(res, "an exec start");
  const body = res.body;
  if (body === null) throw unavailable("Docker sent no output stream");
  const frames = demux(sinks);
  const exit = (async () => {
    for await (const chunk of body) frames.push(chunk);
    frames.end();
    return (await exitCode(engine, id)) ?? noCode();
  })();
  return { exit };
}

File: python/src/threads/adapters/sandboxes/docker/exec.py
"""One Docker exec: created, started, and its multiplexed stream split into the two byte
streams and the exit code the protocol returns (`ExecOutput`).

The stream is the container's own output, so it is a trust boundary: its 8-byte frame headers
are parsed strictly and a frame over the cap is refused, both as a typed `unavailable`.
"""

import asyncio
from collections.abc import Mapping, Sequence
from typing import Final, Literal

from pydantic import JsonValue

from threads.adapters.sandboxes.docker import records
from threads.adapters.sandboxes.docker.engine import Engine, argv_env
from threads.adapters.sandboxes.streams import Pipe, pump
from threads.sandbox.protocol import ExecOutput

HEADER: Final = 8
FRAME_MAX: Final = 64 * 1024 * 1024
_STREAMS: Final[Mapping[int, Literal["stdout", "stderr"]]] = {1: "stdout", 2: "stderr"}
_EXIT_POLLS: Final = 20
_EXIT_PAUSE_S: Final = 0.05

type Chunk = tuple[Literal["stdout", "stderr"], bytes]


class FrameError(Exception):
    """The daemon's multiplexed stream broke its own framing."""


class Frames:
    """Docker's stream multiplexer, fed arbitrary splits: `[stream, 0, 0, 0, len:u32be]`."""

    def __init__(self) -> None:
        self._buf = bytearray()

    def feed(self, chunk: bytes) -> list[Chunk]:
        self._buf += chunk
        out: list[Chunk] = []
        while len(self._buf) >= HEADER:
            head = bytes(self._buf[:HEADER])
            stream = _STREAMS.get(head[0])
            if stream is None or head[1:4] != b"\0\0\0":
                raise FrameError(f"unknown stream frame header {bytes(head).hex()}")
            size = int.from_bytes(head[4:HEADER], "big")
            if size > FRAME_MAX:
                raise FrameError(f"a {size}-byte frame is over the {FRAME_MAX}-byte cap")
            if len(self._buf) < HEADER + size:
                return out
            out.append((stream, bytes(self._buf[HEADER : HEADER + size])))
            del self._buf[: HEADER + size]
        return out

    def end(self) -> None:
        """Raises when the stream stopped inside a frame."""
        if self._buf:
            raise FrameError(f"the stream ended {len(self._buf)} bytes into a frame")


def body(
    cmd: Sequence[str],
    env: Mapping[str, str],
    cwd: str,
    user: str,
) -> Mapping[str, JsonValue]:
    return {
        "AttachStdout": True,
        "AttachStderr": True,
        "AttachStdin": False,
        "Tty": False,
        "User": user,
        "Env": list(argv_env(env)),
        "WorkingDir": cwd,
        "Cmd": list(cmd),
    }


async def start(
    engine: Engine,
    container: str,
    spec: Mapping[str, JsonValue],
    supervised: str | None = None,
) -> ExecOutput:
    """Starts the exec and returns once the daemon accepted it; its output is pumped from the
    stream, which the pump closes.

    `supervised` is the key hash of a command running through the supervisor, whose own exit
    code is in its record rather than in the exec's status (records.py `exit_code`)."""
    made = await engine.exec_create(container, spec)
    stream = await engine.exec_start(made.id)
    pipe = Pipe()

    async def feed(into: Pipe) -> int:
        frames = Frames()
        try:
            async for chunk in stream.aiter_raw():
                for kind, data in frames.feed(chunk):
                    await (into.stdout(data) if kind == "stdout" else into.stderr(data))
            frames.end()
        finally:
            await stream.aclose()
        status = await _exit_code(engine, made.id)
        if supervised is None:
            return status
        state = await records.read_state(engine, container)
        return records.exit_code(state, supervised, status)

    engine.pumps.add(task := pump(pipe, feed))
    task.add_done_callback(engine.pumps.discard)
    return pipe.output()


async def run_to_end(
    engine: Engine, container: str, cmd: Sequence[str], user: str
) -> tuple[int, bytes, bytes]:
    """A short fixed command of the adapter's own (the probe, `--check`): its whole output."""
    output = await start(engine, container, body(cmd, {}, "/", user))
    out = b"".join([c async for c in output.stdout])
    err = b"".join([c async for c in output.stderr])
    return await output.exit_code, out, err


async def _exit_code(engine: Engine, exec_id: str) -> int:
    """The daemon sets `ExitCode` when it reaps the process, which can trail the stream's end
    by a moment."""
    for _ in range(_EXIT_POLLS):
        state = await engine.exec_state(exec_id)
        if not state.running and state.exit_code is not None:
            return state.exit_code
        await asyncio.sleep(_EXIT_PAUSE_S)
    raise FrameError(f"exec {exec_id} reported no exit code after its stream ended")
